src/github.com/t3chguy/riot-static/riot-static.go

```diff
@@ -0,0 +1,234 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+	@TODO Show more than just textual events (Images etc)
+	@TODO Historical member info
+	@TODO make Peeking more useful (paginate etc)
+	@TODO investigate best practices for the spider to get the most out of it
+	@TODO make cache invalidate self after N requests of M minutes
+*/
+
+package main
+
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"github.com/matryer/resync"
+	"net/http"
+	"os"
+	"strconv"
+	"sync"
+)
+
+func paginate(page int, size int, length int) (skip int, end int) {
+	if skip = (page - 1) * size; skip > length {
+		skip = length
+	}
+	if end = skip + size; end > length {
+		end = length
+	}
+	return
+}
+
+func GetPublicRoomsList(c *gin.Context) {
+	var page int
+	var err error
+	if page, err = strconv.Atoi(c.DefaultQuery("page", "1")); err != nil {
+		page = 1
+	}
+
+	pageSize := 20
+
+	data.RLock()
+	skip, end := paginate(page, pageSize, data.NumRooms)
+	c.HTML(http.StatusOK, "rooms.html", gin.H{
+		"Rooms":    data.Ordered[skip:end],
+		"NumRooms": data.NumRooms,
+		"Page":     page,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoom(c *gin.Context) { // ~/room/:roomId/chat
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "room.html", gin.H{
+		"Room": data.Rooms[roomId],
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomServers(c *gin.Context) { // ~/room/:roomId/servers
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "room_servers.html", gin.H{
+		"Room": data.Rooms[roomId],
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMembers(c *gin.Context) { // ~/room/:roomId/members
+	roomId := c.Param("roomId")
+
+	//var page int
+	//var err error
+	//if page, err = strconv.Atoi(c.DefaultQuery("page", "1")); err != nil {
+	//	page = 1
+	//}
+
+	//pageSize := 20
+
+	data.RLock()
+	length := len(data.Rooms[roomId].MemberInfo)
+	//skip, end := paginate(page, pageSize, length)
+	c.HTML(http.StatusOK, "room_members.html", gin.H{
+		"Room": data.Rooms[roomId],
+		//"MemberInfo": data.Rooms[roomId].Members[skip:end],
+		"NumMembers": length,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomPowerLevels(c *gin.Context) { // ~/room/:roomId/power_levels
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "power_levels.html", gin.H{
+		"PowerLevels": data.Rooms[roomId].PowerLevels,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMember(c *gin.Context) { // ~/room/:roomId/members/:mxid
```

Comment:
It's great you've added the path string as a comment like this, but you can do better.  A comment can diverge from the truth.

Why do we need to do this at all? We do this because otherwise lines like `c.Param("mxid")` make no sense at all because the access point here is so far removed from the actual registration of the path. In Dendrite, we get around this by having a tiny in-line function which yanks out the path variables before immediately delegating to another function to do the work: https://github.com/matrix-org/dendrite/blob/master/src/github.com/matrix-org/dendrite/clientapi/routing/routing.go#L83:L88 - you may find it clearer to do something similar.